### PR TITLE
[Golang][client] fix undefined: localVarFile

### PR DESCRIPTION
**Pull request by grokify**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@antihax
@bvwells

### Description of the PR

When building a Go client with a required `formData` parameter of `file` type, a `undefined: localVarFile` error is countered when compiling the client.

This happens because `localVarFile` is currently only defined when a file is optional. This fix does the following:

* defines `localVarFile` when a file is required
* sets `localVarFile` to the variable name as defined in the function call

#### Known Limitations

This approach can only handle a single file, but this is also a limitation of the current Go generator can as `APIClient.prepareRequest()` takes a single pair of `localVarFileName` and `localVarFileBytes` parameters. A future enhancement could support multiple files, either with the same `multipart/form-data` `name` or different ones.

#### Tests

`mvn package` was successfully run in the `samples/client/petstore/go` directory.

Additional Petstore tests were not added for this use case yet because an endpoint with a required `formData` file is not present in the Petstore spec. Adding this endpoint path to the Petstore spec will be a future project.

This has been tested locally with the following Swagger spec with a successful required file upload.

https://github.com/grokify/go-ringcentral/blob/master/codegen/swagger_spec.yaml

#### References

* https://github.com/swagger-api/swagger-codegen/issues/7311
* https://github.com/grokify/go-ringcentral/issues/27

### CODE REVIEW

**Comment by antihax** [comment]
I don't see actual code changes? is this not in scope of the test file?

**Comment by grokify** [comment]
The code changes are in `api.mustache`:

https://github.com/OpenAPITools/openapi-generator/pull/382/files#diff-e70c0e4856318b336d243d2d9a54232f

**Comment by grokify** [comment]
Updates:

* Added `POST /pet/{petId}/uploadImageWithRequiredFile` endpoint to 2.0 and 3.0 `petstore-with-fake-endpoints-models-for-testing.yaml`
* Built new `go-petstore` with `APIClientPetApi.UploadFileWithRequiredFile` function
* Added new test `TestUploadFileRequired` with early return until supported in server. Currently used to ensure client compiles
* Streamlined `api.mustache`

**Comment by wing328** [comment]
@grokify thanks for the fix and the additional test cases to cover the issue moving forward.

Given that there are new test cases, I wonder if you can run the following to update other Petstore samples as well so that the Shippable CI tests will pass:
```
./bin/ruby-petstore.sh > /dev/null 2>&1
./bin/java-petstore-all.sh > /dev/null 2>&1
./bin/java-jaxrs-petstore-server-all.sh > /dev/null 2>&1
./bin/spring-all-pestore.sh > /dev/null 2>&1
./bin/kotlin-client-petstore.sh > /dev/null 2>&1
./bin/kotlin-client-string.sh > /dev/null 2>&1
./bin/kotlin-client-threetenbp.sh > /dev/null 2>&1
./bin/kotlin-server-petstore.shl> /dev/null 2>&1
./bin/php-petstore.sh > /dev/null 2>&1
./bin/php-silex-petstore-server.shj> /dev/null 2>&1
./bin/php-symfony-petstore.sh > /dev/null 2>&1
./bin/php-lumen-petstore-server.sh > /dev/null 2>&1
./bin/php-slim-petstore-server.sh > /dev/null 2>&1
./bin/php-ze-ph-petstore-server.sh > /dev/null 2>&1
./bin/openapi3/php-petstore.sh > /dev/null 2>&1
```

**Comment by grokify** [comment]
Updated the samples and the Shippable CI tests now succeed but others are now failing.

**Circle CI**

> [ERROR] COMPILATION ERROR : 
> [ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/jaxrs-cxf/src/main/java/org/openapitools/api/impl/PetApiServiceImpl.java:[27,8] org.openapitools.api.impl.PetApiServiceImpl is not abstract and does not override abstract method uploadFileWithRequiredFile(java.lang.Long,org.apache.cxf.jaxrs.ext.multipart.Attachment,java.lang.String) in org.openapitools.api.PetApi
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project openapi-cxf-server: Compilation failure
> [ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/jaxrs-cxf/src/main/java/org/openapitools/api/impl/PetApiServiceImpl.java:[27,8] org.openapitools.api.impl.PetApiServiceImpl is not abstract and does not override abstract method uploadFileWithRequiredFile(java.lang.Long,org.apache.cxf.jaxrs.ext.multipart.Attachment,java.lang.String) in org.openapitools.api.PetApi

**Travis CI**

> $ sudo apt-get install -qq bats
> WARNING: The following packages cannot be authenticated!
>   bats
> E: There were unauthenticated packages and -y was used without --allow-unauthenticated

**Comment by antihax** [comment]
@wing328 I ran into similar issues to this before where we wanted to add a test case but doing so broke other clients as they did not support part of the standard.

What would be the best way to proceed in these cases?

**Comment by grokify** [comment]
**CircleCI Error: PetApiServiceImpl is not abstract...**

The following error:

```
PetApiServiceImpl is not abstract and does not override abstract method uploadFileWithRequiredFile
```

is due to `bin/jaxrs-cxf-petstore-server.sh` not updating `samples/server/petstore/jaxrs-cxf/src/main/java/org/openapitools/api/impl/PetApiServiceImpl.java`. Because of this, `samples/server/petstore/jaxrs-cxf/src/main/java/org/openapitools/api/impl/PetApiServiceImpl.java` would not get the new method `uploadFileWithRequiredFile`.

Deleting the `samples/server/petstore/jaxrs-cxf/` directory before running `bin/jaxrs-cxf-petstore-server.sh` resolved this. I have a "test" branch that I did get to successfully build on CircleCI as seen here:

https://circleci.com/gh/grokify/openapi-generator/7

Full error line:

> [ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/jaxrs-cxf/src/main/java/org/openapitools/api/impl/PetApiServiceImpl.java:[27,8] org.openapitools.api.impl.PetApiServiceImpl is not abstract and does not override abstract method uploadFileWithRequiredFile(java.lang.Long,org.apache.cxf.jaxrs.ext.multipart.Attachment,java.lang.String) in org.openapitools.api.PetApi

**Building Samples Not Cleaning Directories**

On another note, it seems `bin/go-petstore.sh` will create new files for `samples/client/petstore/go/go-petstore` but will not remove old files that are no longer generated.

In both cases, it seems like having `bin/jaxrs-cxf-petstore-server.sh` and `bin/go-petstore.sh` remove the target directory would be cleaner. Here's a summary:

| builder | output directory | notes |
|---------|--------------------|--------|
| `bin/jaxrs-cxf-petstore-server.sh` | `samples/server/petstore/jaxrs-cxf/` | Does not overwrite existing file with new method |
| `bin/go-petstore.sh` | `samples/client/petstore/go/go-petstore` | Does not remove old files that are no longer generated |

**Latest CircleCI Error**

The latest CircleCI errors are due to a number of download issues with `scalaz-petstore-client` as listed in the logs.

**Comment by grokify** [comment]
CircleCI [`grokify/openapi-generator/14` fixed](https://circleci.com/gh/grokify/openapi-generator/14). Some sample files like `samples/client/petstore/clojure/pom.xml` are excluded in `.gitignore` and must be preserved.

**Comment by grokify** [comment]
@antihax @bvwells @wing328 

All tests are passing now.

Primary changes are:

* `api.mustache`: https://github.com/OpenAPITools/openapi-generator/pull/382/files#diff-e70c0e4856318b336d243d2d9a54232f
* `2_0/petstore-with-fake-endpoints-models-for-testing.yaml`: https://github.com/OpenAPITools/openapi-generator/pull/382/files#diff-306628117f125fc0db5f1ad40d7b625c
* `3_0/petstore-with-fake-endpoints-models-for-testing.yaml`: https://github.com/OpenAPITools/openapi-generator/pull/382/files#diff-d9b221b1064638087f63507cc10c34c0
* `pet_api_test.go`: https://github.com/OpenAPITools/openapi-generator/pull/382/files#diff-7782decde9080f5a41d4c37aeb579302

File: samples/client/petstore/eiffel/api_client.ecf [hunk 1]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="ISO-8859-1"?>
+<<<<<<< HEAD
```
Comment (jmini) [inline samples/client/petstore/eiffel/api_client.ecf]: Can you please solve this merge conflict?

File: samples/client/petstore/eiffel/test/api_test.ecf [hunk 2]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="ISO-8859-1"?>
+<<<<<<< HEAD
```
Comment (jmini) [inline samples/client/petstore/eiffel/test/api_test.ecf]: Can you please solve this merge conflict?

**Comment by grokify** [comment]
After merging 9b909df543b8c541a17aa5a545473e6e511b140d, Shippable has timed out twice. Any suggestions?

* [Shippable — Run 1085](https://app.shippable.com/github/OpenAPITools/openapi-generator/runs/1085/1/console)
* [Shippable — Run 1084](https://app.shippable.com/github/OpenAPITools/openapi-generator/runs/1084/1/console)

**Comment by grokify** [comment]
All tests are passing with no conflicts.

**Comment by grokify** [comment]
> @wing328 I ran into similar issues to this before where we wanted to add a test case but doing so broke other clients as they did not support part of the standard.
> 
> What would be the best way to proceed in these cases?

@antihax In this case, the samples just had to be rebuilt, but there were some nuances. A PR to enhance `bin/jaxrs-cxf-petstore-server.sh` looks like it may be useful.

If you have this or other test cases to add, feel free to let me know. I'm interested in adding test cases to verify functionality and I'd be happy to look into getting other tests to pass. Doing this a few times can streamline the process.

**Comment by wing328** [comment]
FYI. Moved the new test under `/fake` and make some minor code improvements to Go generator via https://github.com/OpenAPITools/openapi-generator/pull/430

**Comment by jmini** [comment]
Added milestone here

**Comment by grokify** [comment]
> @wing328 I ran into similar issues to this before where we wanted to add a test case but doing so broke other clients as they did not support part of the standard.
> 
> What would be the best way to proceed in these cases?

@antihax Rebuilding the samples should solve these issues, but it may require some debugging. In this particular case, some files that needed to get updated were not getting updated. This required following the log entries which I'll look into providing a PR for.

If you let me know what test case you want to add, I can look into it ensuring tests for other clients pass. I'm generally interested in increasing the number of test cases for better coverage.

**Pull request merged**